shared/AppInsightsCommon/src/Offline.ts

```diff
@@ -0,0 +1,187 @@
+
+    // Set the initial state
+    // rState is changed by the browser, both via events and when we check the navigator.onLine property
+    let rState: eOfflineValue = eOfflineValue.Online;
+    if (_navigator && !isNullOrUndefined(_navigator.onLine) && !_navigator.onLine) { // navigator.onLine is undefined in react-native
+        rState = eOfflineValue.Offline;
+    }
+
+    // ustate is changed by the user calling setOnlineState
+    let uState: eOfflineValue = eOfflineValue.Unknown;
+    // current state would be updated each time rState or uState is changed
+    // it is a resolved value of rState and uState
+    let _currentState: boolean = calCurrentState();
+
+    let _evtNamespace = mergeEvtNamespace(createUniqueNamespace("OfflineListener"), parentEvtNamespace);
+
+    try {
+        if (_enableEvents(getWindow())) {
+            _isListening = true;
+        }
+        
+        if (_document) {
+            // Also attach to the document.body or document
+            let target:any = _document.body || _document;
+
+            if (target.ononline) {
+                if (_enableEvents(target)) {
+                    _isListening = true;
+                }
+            }
+        }
+
+        if (_isListening) {
+            // We are listening to events so lets set the current status rather than assuming we are online #1538
+            if (_navigator && !isNullOrUndefined(_navigator.onLine)) { // navigator.onLine is undefined in react-native
+                _currentState = _navigator.onLine;
+            }
+        }
+    } catch (e) {
+        // this makes react-native less angry
+        _isListening = false;
+    }
+
+    function _enableEvents(target: any): boolean {
+        let enabled = false;
+        if (target) {
+            enabled = eventOn(target, "online", _setOnline, _evtNamespace);
+            if (enabled) {
+                eventOn(target, "offline", _setOffline, _evtNamespace);
+            }
+        }
+        return enabled;
+    }
+
+    function _isOnline(){
+        return _currentState;
+    }
+
+    function calCurrentState(){
+        if (uState === eOfflineValue.Offline || rState === eOfflineValue.Offline){
+            return false;
+        }
+        return true; // if both unknown, then we assume the network is good
+    }
+
+    function listnerNoticeCheck(){
+        // we were offline and are now online or we were online and now offline
+        let newState = calCurrentState();
+        if (_currentState !== newState) {
+            _currentState = newState; // use the resolved state to update
+            // send all the callbacks with the current state
+            arrForEach(listenerList, (callback: OfflineCallback) => {
+                let offlineState: IOfflineState = {
+                    isOnline: _currentState,
+                    rState: rState,
+                    uState: uState
+                };
+                callback(offlineState);
```

Comment:
Lets "protect" this from a faulty callback by wrapping this call in a try / catch so that if one of the callbacks fail then we still call all of the remaining.

```ts
try {
    callback(offlineState);
} catch (e) {
    // Do nothing, just making sure we run all of the callbacks
}
```